crates/bevy_render/src/batching/mod.rs

```diff
@@ -25,65 +29,107 @@ pub struct NoAutomaticBatching;
         };
 
-        // If we are beginning a new batch, record the start item and index
-        if batch_start_item.is_none() {
-            batch_start_item = Some(item);
-            batch_start_index = index.get();
+        let (user_data, buffer_data) = F::get_batch_data(batch_data);
+
+        let buffer_index = gpu_array_buffer.push(buffer_data);
+        *item.batch_range_mut() = buffer_index.index.get()..buffer_index.index.get() + 1;
+        *item.dynamic_offset_mut() = buffer_index.dynamic_offset;
+
+        if no_batching.is_some() {
+            None
+        } else {
+            Some(BatchMeta {
+                pipeline_id: item.cached_pipeline(),
+                draw_function_id: item.draw_function(),
+                dynamic_offset: buffer_index.dynamic_offset,
+                user_data,
+            })
         }
+    };
 
-        if Some(&batch_meta) != next_batch.as_ref().map(|(meta, ..)| meta) {
-            // The next item doesn't match the current batch (or doesn't exist).
-            // Update the phase item to render this batch.
-            let batch_start_item = batch_start_item.take().unwrap();
-            *batch_start_item.batch_range_mut() = batch_start_index..index.get() + 1;
-            *batch_start_item.dynamic_offset_mut() = dynamic_offset;
+    for mut phase in &mut views {
+        let mut items = phase.items.iter_mut().peekable();
+        let mut batch_start_item = None;
+        let mut next_batch = items.peek_mut().and_then(|i| process_item(i));
+        while let Some(item) = items.next() {
+            // Get the current batch meta and update the next batch meta
+            let Some(batch_meta) = std::mem::replace(
+                &mut next_batch,
+                items.peek_mut().and_then(|i| process_item(i)),
+            ) else {
+                // If the current phase item doesn't match the query or has NoAutomaticBatching,
+                // we don't modify it any further
+                continue;
+            };
+
+            let batch_end_item = item.batch_range().end;
+
+            // If we are beginning a new batch, record the start item
+            if batch_start_item.is_none() {
+                batch_start_item = Some(item);
+            }
+
+            if Some(&batch_meta) != next_batch.as_ref() {
+                // The next item doesn't match the current batch (or doesn't exist).
+                // Update the first phase item to render the full batch.
+                let batch_start_item = batch_start_item.take().unwrap();
+                batch_start_item.batch_range_mut().end = batch_end_item;
+            }
         }
     }
 }
+
+pub fn flush_buffer<F: GetBatchData>(
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    gpu_array_buffer: ResMut<GpuArrayBuffer<F::BufferData>>,
+) {
+    let gpu_array_buffer = gpu_array_buffer.into_inner();
+    gpu_array_buffer.write_buffer(&render_device, &render_queue);
+    gpu_array_buffer.clear();
+}
```

Comment:
Good! I wanted to separate this out so we could allow others to append stuff to the buffer for their own purposes.